### ISSUE TITLE
ResourceExhaustedErrorResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[1,100000,512] 

### ISSUE THREAD

**Issue by aclaussen1** [ibab/tensorflow-wavenet#201]
Running on Nvidia geforce 950 2gb memory

I am training on two WAV files of music, both around 23MB.

A single step won't finish. How come if I run on CPU, it can complete steps but not when I run on GPU?

I see similar issues September (https://github.com/ibab/tensorflow-wavenet/issues/10, https://github.com/ibab/tensorflow-wavenet/issues/53, https://github.com/ibab/tensorflow-wavenet/issues/4) but nothing recently.

```
tensorflow/core/common_runtime/bfc_allocator.cc:693]      Summary of in-use Chunks by size: 
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 477 Chunks of size 256 totalling 119.2KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 1024 totalling 4.0KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 1280 totalling 1.2KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 303 Chunks of size 2048 totalling 606.0KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 301 Chunks of size 4096 totalling 1.18MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 149 Chunks of size 32768 totalling 4.66MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 40960 totalling 40.0KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 61440 totalling 60.0KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 65536 totalling 256.0KiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 524288 totalling 2.00MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 1048576 totalling 4.00MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 52 Chunks of size 6400000 totalling 317.38MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 9 Chunks of size 6400256 totalling 54.93MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 6400512 totalling 18.31MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 6401024 totalling 18.31MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 5 Chunks of size 6402048 totalling 30.53MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 6404096 totalling 6.11MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 6420480 totalling 6.12MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 6422528 totalling 24.50MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 6455296 totalling 24.62MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 12791040 totalling 12.20MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 12798208 totalling 12.21MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 12800000 totalling 12.21MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 4 Chunks of size 12800256 totalling 48.83MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 12800512 totalling 36.62MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 12801024 totalling 36.62MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 12802048 totalling 36.63MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 12804096 totalling 36.63MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 12812288 totalling 36.66MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 3 Chunks of size 12828672 totalling 36.70MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 18937344 totalling 18.06MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 2 Chunks of size 19220480 totalling 36.66MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 2 Chunks of size 19245056 totalling 36.71MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 25434624 totalling 24.26MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 102400000 totalling 97.66MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 102401024 totalling 97.66MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 204800000 totalling 195.31MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:696] 1 Chunks of size 282473472 totalling 269.39MiB
I tensorflow/core/common_runtime/bfc_allocator.cc:700] Sum Total of in-use chunks: 1.56GiB
I tensorflow/core/common_runtime/bfc_allocator.cc:702] Stats: 
Limit:                  1686568960
InUse:                  1672186112
MaxInUse:               1684970240
NumAllocs:                    2706
MaxAllocSize:            365845504

W tensorflow/core/common_runtime/bfc_allocator.cc:274] ************************************************************************************************xxxx
W tensorflow/core/common_runtime/bfc_allocator.cc:275] Ran out of memory trying to allocate 195.31MiB.  See logs for memory state.
W tensorflow/core/framework/op_kernel.cc:975] Resource exhausted: OOM when allocating tensor with shape[1,100000,512]
Storing checkpoint to ./logdir/train/2017-01-04T03-21-56 ... Done.
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1021, in _do_call
    return fn(*args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1003, in _run_fn
    status, run_metadata)
  File "/usr/lib/python3.5/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/errors_impl.py", line 469, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[1,100000,512]
	 [[Node: wavenet_1/postprocessing/add_18 = Add[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"](wavenet_1/postprocessing/add_17, wavenet_1/dilated_stack/layer18/add_1)]]
	 [[Node: wavenet_1/loss/Mean/_13 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_12278_wavenet_1/loss/Mean", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "train.py", line 316, in <module>
    main()
  File "train.py", line 293, in main
    summary, loss_value, _ = sess.run([summaries, loss, optim])
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 766, in run
    run_metadata_ptr)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 964, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1014, in _do_run
    target_list, options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1034, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[1,100000,512]
	 [[Node: wavenet_1/postprocessing/add_18 = Add[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"](wavenet_1/postprocessing/add_17, wavenet_1/dilated_stack/layer18/add_1)]]
	 [[Node: wavenet_1/loss/Mean/_13 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_12278_wavenet_1/loss/Mean", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

Caused by op 'wavenet_1/postprocessing/add_18', defined at:
  File "train.py", line 316, in <module>
    main()
  File "train.py", line 234, in main
    loss = net.loss(audio_batch, args.l2_regularization_strength)
  File "/home/alex/Downloads/wavenet/tensorflow-wavenet-master/wavenet/model.py", line 491, in loss
    raw_output = self._create_network(network_input)
  File "/home/alex/Downloads/wavenet/tensorflow-wavenet-master/wavenet/model.py", line 326, in _create_network
    total = sum(outputs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/math_ops.py", line 814, in binary_op_wrapper
    return func(x, y, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_math_ops.py", line 73, in add
    result = _op_def_lib.apply_op("Add", x=x, y=y, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 759, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 2240, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1128, in __init__
    self._traceback = _extract_stack()

ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[1,100000,512]
	 [[Node: wavenet_1/postprocessing/add_18 = Add[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"](wavenet_1/postprocessing/add_17, wavenet_1/dilated_stack/layer18/add_1)]]
	 [[Node: wavenet_1/loss/Mean/_13 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_12278_wavenet_1/loss/Mean", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]


```

**Comment by akademi4eg** [comment]
@aclaussen1 I think 2Gb is not enough to store model parameters and one train batch. The shape [1,100000,512] is probably allocated for input of one of layers. My GPU has 8Gb and around 7.5Gb are allocated during training.
You can try decreasing --sample_size parameter. It regulates number of samples in one batch. Default value is 100000. Try 25000 or even 10000.

**Comment by uralik** [comment]
@akademi4eg probably you see 7.5G allocated just due to the TF policy - it really likes to allocate a lot of room for its process. (in comparison to modest Theano). But actually this policy allows to solve problem with GPU memory fragmentation during the run-time.

@aclaussen1 bfc_allocator definitely means that your gpu is out of memory. As @akademi4eg mentioned - you could decrease the sample size OR number of channels in the dilated stack OR number of stacks.

**Comment by mikelibg** [comment]
Getting same OOM when running on 61GiB aws instance even with sample size 10,000.
commit: 3c973c038c8c2c20fef0039f111cb04139ff594b
python: 3.6
with --silence_threshold=0 on one utterance p225 from the original VCTK corpus

```
>>> psutil.virtual_memory()
svmem(total=64389132288, available=60933558272, percent=5.4, used=2929561600, free=59257782272, active=3300839424, inactive=1334075392, buffers=73183232, cached=2128605184, shared=21499904, slab=143368192)
```

Exception:
2018-05-13 06:20:26.790439: W tensorflow/core/framework/op_kernel.cc:1318] OP_REQUIRES failed at slice_op.cc:154 : Resource exhausted: OOM when allocating tensor with shape[1,44050,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
Storing checkpoint to ./logdir/train/2018-05-13T06-19-22 ... Done.
Traceback (most recent call last):
  File "/home/ubuntu/training/tensorflow-wavenet/venv/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1322, in _do_call
    return fn(*args)
  File "/home/ubuntu/training/tensorflow-wavenet/venv/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1307, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/home/ubuntu/training/tensorflow-wavenet/venv/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1409, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.ResourceExhaustedError: OOM when allocating tensor with shape[1,38934,256] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[Node: wavenet_1/loss/Slice = Slice[Index=DT_INT32, T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"](wavenet_1/loss/Reshape, wavenet_1/loss/Slice/begin, wavenet_1/loss/Slice/size)]]
